### DISCUSSION TITLE
Long document context not working

### DISCUSSION THREAD

**Discussion by NotADeveloperAtAll** [open-webui/open-webui#749] [General]
I'm trying to chat with documents. It works pretty well as long as the text isn't too long. If the text is too long, the model struggles with accuracy and finding the data.  I've tried multiple file formats.

What is going on? 
Are there ways to go around this?
Is there a limit in the amount of characters for instance?

**Comment by justinh-rahb** [comment]
Depends on the model. Mistral supports up to 32k for example, whereas llama2 tops out at 4096 tokens. A general rule of thumb is 100 tokens is around 75 words.

**Reply by NotADeveloperAtAll** [reply]
Thanks.
The document is around 10k tokens. Even with Mistral, it struggles to find what I ask.

**Reply by justinh-rahb** [reply]
This could be down to how the current RAG implementation has been tuned. Unfortunately there's not a great "one-size-fits-all" approach to that AFAIK. Would be worth seeing if any alternative projects with RAG such as [PrivateGPT](https://github.com/imartinez/privateGPT) exhibit similar issues for your document.

**Reply by tdiprima** [reply]
@justinh-rahb  Could you please clarify how the current RAG implementation works? I'm working with a collection of over 1,000 short text documents (each just a single paragraph). When I select the collection and ask a question, the response only includes 3 documents. I expected more relevant documents in the results.

It seems like the model is only retrieving or "seeing" those 3 specific documents, rather than accessing the full set of 1,032. Without visibility into the remaining 1,029, it's unable to provide a more comprehensive answer or additional document references related to my query. Any insights on how this retrieval is configured?

**Reply by pbpippen** [reply]
It uses the most relevant documents based on your document Query Params. Your top K in document settings is probably set as 3. You may want to either increase it slightly, enable hybrid search, or use different embedding and reranking models

**Reply by tdiprima** [reply]
@pbpippen Thank you for the quick response!  You were absolutely right—the "top K" was set to 3, and increasing it resolved the issue.  I really appreciate the guidance.

I'm curious about the other suggestions you mentioned—specifically, "enabling hybrid search" and "using different embedding and reranking models." Could you explain a bit more about what these mean or share any examples?  I'm not too familiar with those options and would love to understand how they might improve the retrieval process.

Thanks again for your help!

**Reply by pbpippen** [reply]
Hybrid search combines vector search and keyword search to try to get most relevant documents.
The embedding model helps understand the meaning of words in documents, while the re-ranker ensures the most relevant documents are prioritized after retrieval. By using good quality models for both, your RAG will perform better.
